import { ResponseError } from '@audius/sdk'

type ErrorWithMessage = {
  name: string
  message: string
}

function isErrorWithMessage(error: unknown): error is ErrorWithMessage {
  return (
    typeof error === 'object' &&
    error !== null &&
    'message' in error &&
    typeof (error as Record<string, unknown>).message === 'string'
  )
}

export function toErrorWithMessage(maybeError: unknown): ErrorWithMessage {
  if (isErrorWithMessage(maybeError)) return maybeError

  try {
    return new Error(JSON.stringify(maybeError))
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError))
  }
}

export function getErrorMessage(error: unknown) {
  return toErrorWithMessage(error).message
}

export const isResponseError = (error: unknown): error is ResponseError =>
  error instanceof Error && error.name === 'ResponseError'

export class ErrorWithCause extends Error {
  cause: unknown

  constructor(message: string, cause?: unknown) {
    super(message)
    this.name = 'ErrorWithCause'
    this.cause = cause

    Object.setPrototypeOf(this, ErrorWithCause.prototype)
  }
}
